core/test/element_util_test.cpp

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+
+#include "scipp/core/element/util.h"
+#include "scipp/units/unit.h"
+
+#include "fix_typed_test_suite_warnings.h"
+
+#include <cmath>
+#include <gtest/gtest.h>
+#include <limits>
+#include <vector>
+
+using namespace scipp::core::element;
+
+TEST(ElementUtilTest, convertMaskedToZero_masks_special_vals) {
+  const std::vector<double> x{1.0, std::numeric_limits<double>::quiet_NaN(),
+                              std::numeric_limits<double>::infinity(), 1.0};
+  const std::vector<bool> masks{true, true, true, false};
+
+  // Should not replace values which are not masked
+  const std::vector<double> expected{0, 0, 0, 1.0};
+
+  assert(x.size() == masks.size());
+  assert(x.size() == expected.size());
+
+  for (size_t i = 0; i < x.size(); i++) {
+    EXPECT_EQ(convertMaskedToZero(x[i], masks[i]), expected[i]);
+  }
+}
+
+TEST(ElementUtilTest, convertMaskedToZero_ignores_unmasked) {
+  EXPECT_TRUE(std::isnan(
+      convertMaskedToZero(std::numeric_limits<double>::quiet_NaN(), false)));
+
+  EXPECT_TRUE(std::isinf(
+      convertMaskedToZero(std::numeric_limits<double>::infinity(), false)));
+}
+
+TEST(ElementUtilTest, convertMaskedToZero_forwards_units) {
+  const auto dimensionless = scipp::units::dimensionless;
+
+  const std::vector<scipp::units::Unit> inputUnits{
+      scipp::units::m, scipp::units::dimensionless, scipp::units::s};
+
+  for (const auto &unit : inputUnits) {
+    // Unit 'a' should always be preserved
+    EXPECT_EQ(convertMaskedToZero(unit, dimensionless), unit);
+  }
+}
+
+TEST(ElementUtilTest, convertMaskedToZero_accepts_all_types) {
+
+  auto testType = [](const auto &val) {
+    EXPECT_EQ(convertMaskedToZero(val, true), decltype(val){0});
```

Comment:
In this case it's to check that the operator accepts the named types in the parameter pack, since I originally had linker errors (which were much harder to track down) as we didn't support `int64_t`

What about splitting it into two so it's obvious were checking first, that it accepts T and second that T does not change like so?

```
EXPECT_EQ(convertMaskedToZero(val, true), 0);
EXPECT_TRUE(std::is_same_v<decltype(convertMaskedToZero(val, true)), decltype(val)>);
```

Reply:
Edit: I've used a static assert to the same effect which should make it clear what the intention is